REST API: Add Field Controller

mdawaffe opened the pull request:
Just as `WP_REST_Controller` is mostly just a fancy wrapper for
`register_rest_route()`, `WPCOM_REST_API_V2_Field_Controller` is a wrapper
for `register_rest_field()`.

#### Changes proposed in this Pull Request:

The new class handles a few conveniences:

* Returning the correct things (errors, default values, etc.) for
  permission check errors.
* Ensuring output is always correctly typecast.
* `?context=...` filtering for nested structures, which core does not
  handle correctly.

#### Testing instructions:
`yarn docker:phpunit --group rest-api`

#### Proposed changelog entry for your changes:

None required.

mdawaffe committed:
REST API: Add Field Controller

Just as `WP_REST_Controller` is mostly just a fancy wrapper for
`register_rest_route()`, `WPCOM_REST_API_V2_Field_Controller` is a wrapper
for `register_rest_field()`. It handles a few conveniences:
* Returning the correct things (errors, default values, etc.) for
  permission check errors.
* Ensuring output is always correctly typecast.
* `?context=...` filtering for nested structures, which core does not
  handle correctly.

ockham commented on _inc/lib/core-api/class-wpcom-rest-field-controller.php:
```diff
@@ -0,0 +1,326 @@
+	 * Permission Check for the field's getter. Must be implemented in the inheriting class.
+	 *
+	 * @param mixed $object_data Whatever the endpoint would returnn for its response.
+	 * @param WP_REST_Request
```
Should we specify the param name (`$request`) here? (Same for the next couple of methods below.)

ockham reviewed (dismissed):
Left one minor [note](https://github.com/Automattic/jetpack/pull/10600/files#r232834985). LGTM otherwise! :ship:

tyxla commented on _inc/lib/core-api/class-wpcom-rest-field-controller.php:
```diff
@@ -0,0 +1,326 @@
+	/**
+	 * Permission Check for the field's getter. Must be implemented in the inheriting class.
+	 *
+	 * @param mixed $object_data Whatever the endpoint would returnn for its response.
```
Typo: `returnn`; it also occurs a couple more times.

tyxla reviewed (dismissed):
This is looking and working great for me 👍 💯 

I've left some minor suggestions, nothing blocking though, so feel free to take or leave any of them.

Feel free to 🚢 at will!

mdawaffe committed:
REST API Field Controller: Typos in PHPDoc

mdawaffe committed:
REST API Field Controller: Code Standards

mdawaffe committed:
REST API Field Controller: Translation fixes.

mdawaffe committed:
REST API Field Controller: Warn that returning `false` from is incorrect.

mdawaffe committed:
REST API Field Controller: Typos

mdawaffe committed:
REST API Field Controller: Method visibility

mdawaffe commented on _inc/lib/core-api/class-wpcom-rest-field-controller.php:
```diff
@@ -0,0 +1,326 @@
+	 * @param WP_REST_Request $request
+	 * @return mixed
+	 */
+	function prepare_for_response( $value, $request ) {
```
Fixed in dc5ec045d.

tyxla reviewed (dismissed):
This is looking great, tests are passing too 👍

🚢

lezama commented:
🚢 :shipit:

mdawaffe committed:
REST API Field Controller: Linter fixes

lezama commented on _inc/lib/core-api/class-wpcom-rest-field-controller.php:
```diff
@@ -38,24 +38,28 @@ public function __construct() {
 	 * Ensures the response matches the schema and request context.
 	 *
-	 * @param mixed $value
+	 * @param mixed           $value
```
is the big `           ` expected?

mdawaffe replied:
Yes.

lezama commented on _inc/lib/core-api/class-wpcom-rest-field-controller.php:
```diff
@@ -161,55 +165,56 @@ private function prepare_for_response( $value, $request ) {
 	public function get_permission_check( $object_data, $request ) {
 		/* translators: %s: get_permission_check() */
-                _doing_it_wrong( 'WPCOM_REST_API_V2_Field_Controller::get_permission_check', sprintf( __( "Method '%s' must be overridden.", 'jetpack' ), __METHOD__ ), 'Jetpack 6.8' );
+				_doing_it_wrong( 'WPCOM_REST_API_V2_Field_Controller::get_permission_check', sprintf( __( "Method '%s' must be overridden.", 'jetpack' ), __METHOD__ ), 'Jetpack 6.8' );
```
is this linting ok?

mdawaffe replied:
no - looks like it got confused about some spaces. I'll fix.

mdawaffe replied:
Thanks for noticing! Fixed in 6a22f1148.

mdawaffe committed:
REST API Field Controller: Linter fixes fixes

lezama reviewed (dismissed):
After Linter fixes fixes let me approve this again

oskosk commented on _inc/lib/core-api/class-wpcom-rest-field-controller.php:
```diff
@@ -0,0 +1,326 @@
+	final public function update_from_request( $value, $object_data, $field_name, $request, $object_type ) {
+		$permission_check = $this->update_permission_check( $value, $object_data, $request );
+
+		if ( ! $permission_check ) {
```
I concur

mdawaffe replied:
Fixed in 0b6042f3a.

mdawaffe committed:
REST API Field Controller: Normalize `update_from_request()` return value

Return `WP_Error` from `update_from_request()` when `update_permission_check()` returns false.

zinigor reviewed (approved):
Looks great, thanks!
